File: code-coaches/medium/camel-to-snake/CamelToSnake.java
import java.util.Scanner;

public class CamelToSnake {

    public static String toSnakeCase(String camelCaseString) {
        String snakeCaseString = "";

        for (int i = 0; i < camelCaseString.length(); i++) {
            char currentChar = camelCaseString.charAt(i);

            if (i != 0) {
                char previousChar = camelCaseString.charAt(i - 1);

                if ((Character.isUpperCase(currentChar) && Character.isLetter(previousChar))
                    || (Character.isLetter(currentChar) && Character.isDigit(previousChar)))
                {
                    snakeCaseString += "_";
                }
            }

            snakeCaseString += Character.toLowerCase(currentChar);
        }

        return snakeCaseString;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String camelCaseString = scanner.nextLine();

        System.out.println(toSnakeCase(camelCaseString));

        scanner.close();
    }
}

File: code-coaches/medium/camel-to-snake/CamelToSnake.cs
using System;

public class CamelToSnake
{
    static string ToSnakeCase(string camelCaseString)
    {
        string snakeCaseString = "";

        for (int i = 0; i < camelCaseString.Length; i++)
        {
            char currentChar = camelCaseString[i];

            if (i != 0)
            {
                char previousChar = camelCaseString[i - 1];

                if ((char.IsUpper(currentChar) && char.IsLetter(previousChar))
                    || (char.IsLetter(currentChar) && char.IsDigit(previousChar)))
                {
                    snakeCaseString += "_";
                }
            }

            snakeCaseString += char.ToLower(currentChar);
        }

        return snakeCaseString;
    }

    static void Main(string[] args)
    {
        string camelCaseString = Console.ReadLine();
        Console.WriteLine(ToSnakeCase(camelCaseString));
    }
}
